### PR TITLE
feat(twap): display twap-specific information on order receipt modal

### LINKED ISSUE #2627: TWAP order receipt modal
Figma: https://www.figma.com/file/vfHjG18yxhe1SnNMhXy9f5/TWAP-Orders-UI?type=design&node-id=466-4354&t=Rf3UuavsHvo92yk0-0

image

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Fixes #2627

## Out of scope
1. Filled %, surplus, fee - these values are hidden for parent TWAP orders, will be fixed in https://github.com/cowprotocol/cowswap/issues/2640
2. Order Id and link to Explorer are hidden for parent TWAP orders because currently Explorer doesn't support TWAP orders

Parent:
image

Children:
image

image

  # To Test
1. Displayed the number of parts on the top of the modal. Important: the text is not aligned to the center as in Figma, because I've used the standard `InlineBanner` component. I think we should be consistent
2. Displayed a hash and a link to the Safe transaction
3. Displayed Safe tx nonce
4. Displayed a count of Safe tx confirmations

### CODE REVIEW

File: src/common/pure/GnosisSafeLink/index.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { ExternalLink } from 'legacy/theme'
+
+import { getSafeWebUrl } from 'api/gnosisSafe'
+
+export function GnosisSafeLink(props: {
```
Comment (alfetopito) [inline src/common/pure/GnosisSafeLink/index.tsx:5]: It's no longer called `Gnosis Safe`, can you rename it?

I was thinking to name it `SafeLink`, but it's too ambiguous.
`SafeWalletLink` is more precise IMO

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx [hunk 2]
```diff
@@ -82,6 +88,12 @@ export function ReceiptModal({
           <CloseIcon onClick={() => onDismiss()} />
         </styledEl.Header>
 
+        {twapOrder && (
+          <styledEl.InfoBannerWrapper>
+            <InlineBanner type="information" content={`Part of a ${twapOrder.order.n}-part TWAP order split`} />
```
Comment (alfetopito) [inline src/modules/ordersTable/pure/ReceiptModal/index.tsx:93]: Could we mention which part this is?

Like: `Part 5 of a 10 part TWAP...`

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx [hunk 3]
```diff
@@ -144,14 +156,27 @@ export function ReceiptModal({
...
+              <styledEl.Field>
+                {order.executionData.activityId && (
+                  <>
+                    <FieldLabel label={order.executionData.activityTitle} />
+                    <IdField id={order.executionData.activityId} chainId={chainId} />
+                  </>
+                )}
+              </styledEl.Field>
+            )}
+            {twapOrder?.safeParams && (
```
Comment (alfetopito) [inline src/modules/ordersTable/pure/ReceiptModal/index.tsx:170]: Can you make the safe params part more generic?

The goal is to show this also for non-TWAP but still Safe orders

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx (hunk 2 above)
Comment (alfetopito) [reply src/modules/ordersTable/pure/ReceiptModal/index.tsx:93]: Ohhhh.
Then it's not a part of. It is THE order itself

File: src/common/pure/GnosisSafeLink/index.tsx (hunk 1 above)
Comment (shoom3301) [reply src/common/pure/GnosisSafeLink/index.tsx:5]: Fair enough, renamed

File: src/legacy/assets/cow-swap/safe-logo.svg
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/legacy/assets/cow-swap/safe-logo.svg:1]: I haven't found an existing one.
Moved to the suggested directory

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx (hunk 2 above)
Comment (shoom3301) [reply src/modules/ordersTable/pure/ReceiptModal/index.tsx:93]: I'm confused you, sorry. Then the text should be like:
```
TWAP order split into 5 parts
```

I will rework it

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx (hunk 3 above)
Comment (shoom3301) [reply src/modules/ordersTable/pure/ReceiptModal/index.tsx]: Make sense. Refactored:
https://github.com/cowprotocol/cowswap/pull/2652/commits/4b8849976697861431e828f90b806447f6fe1daf

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx (hunk 2 above)
Comment (shoom3301) [reply src/modules/ordersTable/pure/ReceiptModal/index.tsx:94]: Parent:
image

Children:
image

>Could we mention which part this is?

Unfortunately, we can not right now. It's a part of the task: https://github.com/cowprotocol/cowswap/issues/2640

Comment (shoom3301) [reply src/modules/ordersTable/pure/ReceiptModal/index.tsx]: Fixed:
https://github.com/cowprotocol/cowswap/pull/2652/commits/07316dbd8c0cce8055066d2215e2daf2bf10031e

File: src/modules/twap/hooks/useTwapOrderById.ts
Comment (alfetopito) [inline src/modules/twap/hooks/useTwapOrderById.ts:1]: Similar to the helper fn from another PR, can you renamed it to be explicit:

`useTwapOrderById` -> `useTwapOrderByParentId`

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx [hunk 4]
```diff
@@ -82,6 +91,19 @@ export function ReceiptModal({
...
+            <InlineBanner
+              type="information"
+              content={
+                isTwapPartOrder
+                  ? `Part of a ${twapOrder.order.n}-part TWAP order split`
+                  : `TWAP order split into ${twapOrder.order.n} parts`
+              }
+            />
+          </styledEl.InfoBannerWrapper>
+        )}
```
Comment (alfetopito) [inline src/modules/ordersTable/pure/ReceiptModal/index.tsx:105]: Not for this PR, just realized this now.

Isn't this violating the isolation principle?
The ordersTable component knows about specific order types.
Maybe it should instead be like TradeWidget and similar where the custom fields are consumed as slots?

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx [hunk 5]
```diff
@@ -114,20 +136,26 @@ export function ReceiptModal({
...
-              <SurplusField order={order} />
-            </styledEl.Field>
-
-            <styledEl.Field>
-              <FieldLabel label="Fee" tooltip={tooltips.FEE} />
-              <FeeField order={order} />
-            </styledEl.Field>
+            {/*TODO: Currently, we don't have this information for parent TWAP orders*/}
+            {/*The condition should be removed once we have the data*/}
+            {(!twapOrder || isTwapPartOrder) && (
```
Comment (alfetopito) [inline src/modules/ordersTable/pure/ReceiptModal/index.tsx:141]: One way to keep the receipt modal agnostic is to have a prop to toggle these optional fields. What do you think?

Also not mandatory for this PR, just raising like the previous comment

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx [hunk 6]
```diff
@@ -144,14 +172,27 @@ export function ReceiptModal({
...
+              <styledEl.Field>
+                {order.executionData.activityId && (
+                  <>
+                    <FieldLabel label={order.executionData.activityTitle} />
+                    <IdField id={order.executionData.activityId} chainId={chainId} />
+                  </>
+                )}
+              </styledEl.Field>
+            )}
+            {twapOrder && safeTxParams && (
```
Comment (alfetopito) [inline src/modules/ordersTable/pure/ReceiptModal/index.tsx:186]: Can this be based only on `safeTxParams` instead?
This still requires twapOrder, but should be usable by SWAP and LIMIT orders as well

**Review by alfetopito — APPROVED** [review]
Seems to work

Screenshot 2023-06-14 at 15 44 10

Just didn't get the part order created to check that, though...

As for the code, a couple of comments, not blocking, probably different issues/PRs

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx (hunk 4 above)
Comment (shoom3301) [reply src/modules/ordersTable/pure/ReceiptModal/index.tsx:105]: It's definitely should, my bad :(
Just don't have enough time, to refactor this module and add slots

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx (hunk 5 above)
Comment (shoom3301) [reply src/modules/ordersTable/pure/ReceiptModal/index.tsx:141]: 100% agree

File: src/modules/ordersTable/pure/ReceiptModal/index.tsx (hunk 4 above)
Comment (alfetopito) [reply src/modules/ordersTable/pure/ReceiptModal/index.tsx:105]: Created an issue so we don't forget https://github.com/cowprotocol/cowswap/issues/2671

**Pull request merged**